### PR TITLE
[GSDESCARTE-17] Map size in real world units

### PR DESCRIPTION
**Pull request by tkohr**
PR allows to specify size of printed map canvas in "in", "cm", "mm" and "m" as it is already documented in readme (`"size": [13, 13, "cm"]`).

### CODE REVIEW

File: src/printer/job.js [hunk 1]
```diff
@@ -126,3 +128,41 @@ function getFrameState(spec) {
 
   return frameState;
 }
+
+/**
+ * Returns the map canvas size in pixels based on size units and dpi given in spec
+ * @param {PrintSpec} spec
+ * @return {Array}
```
Comment (fgravin) [inline src/printer/job.js:135]: array of what ?

File: src/printer/job.js [hunk 2]
```diff
@@ -20,7 +20,8 @@ let counter = 0;
  * @param {PrintSpec} spec
  */
 export function createJob(spec) {
-  const frameState = getFrameState(spec);
+  const pixelSize = calculatePixelSize(spec);
```
Comment (fgravin) [inline src/printer/job.js:23]: I think the name is confusing but maybe it's me. I understand it is the size of a pixel, but it is actually the size in pixel units.

File: src/printer/job.js [hunk 3]
```diff
@@ -126,3 +128,41 @@ function getFrameState(spec) {
...
+  let pixelY;
+  const unit = spec.size[2];
+
+  switch (unit) {
+    case 'in':
+      pixelX = spec.dpi * spec.size[0];
+      pixelY = spec.dpi * spec.size[1];
+      break;
+    case 'cm':
+      pixelX = (spec.dpi * spec.size[0]) / 2.54;
```
Comment (fgravin) [inline src/printer/job.js:151]: create a const for `2.54`, that could be global.

File: src/printer/job.js [hunk 4]
```diff
@@ -126,3 +128,41 @@ function getFrameState(spec) {
...
   return frameState;
 }
+
+/**
+ * Returns the map canvas size in pixels based on size units and dpi given in spec
+ * @param {PrintSpec} spec
+ * @return {Array}
+ */
+function calculatePixelSize(spec) {
+  if (!spec.size[2] || spec.size[2] === 'px') {
```
Comment (fgravin) [inline src/printer/job.js:138]: use destruct to have `dpi` and `size` const.

File: src/printer/job.js [hunk 5]
```diff
@@ -126,3 +128,41 @@ function getFrameState(spec) {
 
   return frameState;
 }
+
+/**
+ * Returns the map canvas size in pixels based on size units and dpi given in spec
+ * @param {PrintSpec} spec
+ * @return {Array}
+ */
+function calculatePixelSize(spec) {
```
Comment (fgravin) [inline src/printer/job.js:137]: All your code should be tested.
Add a test for this function please

**Review by fgravin — CHANGES_REQUESTED** [review]
Great work, the logic is good.
Some more work would be appreciate, slight refactoring and test, you must test all the function you create, thanks.

File: src/printer/job.js (hunk 4 above)
Comment (tkohr) [reply src/printer/job.js:138]: not sure what you mean here

File: src/printer/job.js (hunk 2 above)
Comment (tkohr) [reply src/printer/job.js:23]: ok changed it to `sizeInPixel` ;-)

File: src/printer/job.js (hunk 1 above)
Comment (tkohr) [reply src/printer/job.js:135]: ah yep `Array<Number>`

File: src/printer/job.js (hunk 4 above)
Comment (fgravin) [reply src/printer/job.js:138]:
```js
const { size, dpi } = spec
```

Comment (tkohr) [reply src/printer/job.js]: ah cool wasn't aware of that, thanks

File: src/printer/job.js (hunk 5 above)
Comment (tkohr) [reply src/printer/job.js]: ok I added one test, thanks for your help

**Comment by tkohr** [comment]
@fgravin thanks for the review & help

branch is adapted, rebased and fixes rebase from prior PR (oups) ;-)

File: test/unit/printer/job.test.js [hunk 6]
```diff
@@ -53,6 +61,11 @@ describe('job creation', () => {
   it('creates the correct amount of layers', () => {
     expect(LayersMock.createLayer).toHaveBeenCalledTimes(3);
   });
+  it('creates the correct canvas size from cm', () => {
+    let xPixelExpected = Math.round(spec.dpi * spec.size[0] / METRIC_DENOMINATOR);
```
Comment (fgravin) [inline test/unit/printer/job.test.js:65]: I think test should expect concrete value sand not expression or functions that comes from the tested function.

File: src/shared/constants.js [hunk 7]
```diff
@@ -1,2 +1,3 @@
 export const MESSAGE_JOB_REQUEST = 'requestJob';
 export const MESSAGE_JOB_STATUS = 'jobStatus';
+export const METRIC_DENOMINATOR = 2.54;
```
Comment (fgravin) [inline src/shared/constants.js:3]: 2.54 is the amount of inches in one centimer, not sure the name is releavant.

File: test/unit/printer/job.test.js (hunk 6 above)
Comment (tkohr) [reply test/unit/printer/job.test.js]: ok I changed that

File: src/shared/constants.js (hunk 7 above)
Comment (tkohr) [reply src/shared/constants.js]: ok I renamed it

**Review by fgravin — APPROVED** [review]

File: src/printer/job.js [hunk 8]
```diff
@@ -80,9 +82,10 @@ export function createJob(spec) {
 /**
  * Returns an OpenLayers frame state for a given job spec
  * @param {PrintSpec} spec
+ * @param {Array} pixelSize
  * @return {FrameState}
  */
-function getFrameState(spec) {
+function getFrameState(spec, pixelSize) {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/printer/job.js]: at least one is keeping his eyes open ;-) thanks!

**Pull request merged**